introduce new release process

garyverhaegen-da opened the pull request:
Context
=======

After multiple discussions about our current release schedule and process, we've come to the conclusion that we need to be able to make a distinction between technical snapshots and marketing releases. In other words, we need to be able to create a bundle for early adopters to test without making it an officially-supported version, and without necessarily implying everyone should go through the trouble of upgrading. The underlying goal is to have less frequent but more stable "official" releases.

This PR is a proposal for a new release process designed under the following constraints:

- Reuse as much as possible of the existing infrastructure, to minimize effort but also chances of disruptions.
- Have the ability to create "snapshot"/"nightly"/... releases that are not meant for general public consumption, but can still be used by savvy users without jumping through too many extra hoops (ideally just swapping in a slightly-weirder version string).
- Have the ability to promote an existing snapshot release to "official" release status, with as few changes as possible in-between, so we can be confident that the official release is what we tested as a prerelease.
- Have as much of the release pipeline shared between the two types of releases, to avoid discovering non-transient problems while trying to promote a snapshot to an official release.
- Triggerring a release should still be done through a PR, so we can keep the same approval process for SOC2 auditability.

The gist of this proposal is to replace the current `VERSION` file with a `LATEST` file, which would have the following format:

```
ef5d32b7438e481de0235c5538aedab419682388 0.13.53-alpha.20200214.3025.ef5d32b7
```

This file would be maintained with a script to reduce manual labor in producing the version string. Other than that, the process will be largely the same, with releases triggered by changes to this `LATEST` and the release notes files.

Version numbers
===============

Because one of the goals is to reduce the velocity of our published version numbers, we need a different version scheme for our snapshot releases. Fortunately, most version schemes have some support for that; unfortunately, the SDK sits at the intersection of three different version schemes that have made incompatible choices. Without going into too much detail:

- Semantic versioning (which we chose as the version format for the SDK version number) allows for "prerelease" version numbers as well as "metadata"; an example of a complete version string would be `1.2.3-nightly.201+server12.43`. The "main" part of the version string always has to have 3 numbers separated by dots; the "prerelease" (after the `-` but before the `+`) and the "metadata" (after the `+`) parts are optional and, if present, must consist of one or more segments separated by dots, where a segment can be either a number or an alphanumeric string. In terms of oredering, metadata is irrelevant and any version with a prerelease string is before the corresponding "main" version string alone. Amongst prereleases, segments are compared in order with purely numeric ones compared as numbers and mixed ones compared lexicographically. So 1.2.3 is more recent than 1.2.3-1, which is itself less recent than 1.2.3-2.
- Maven version strings are any number of segments separated by a `.`, a `-`, or a transition between a number and a letter. Version strings are compared element-wise, with numeric segments being compared as numbers. Alphabetic segments are treated specially if they happen to be one of a handful of magic words (such as "alpha", "beta" or "snapshot" for example) which count as "qualifiers"; a version string with a qualifier is "before" its prefix (`1.2.3` is before `1.2.3-alpha.3`, which is the same as `1.2.3-alpha3` or `1.2.3-alpha-3`), and there is a special ordering amongst qualifiers. Other alphabetic segments are compared alphabetically and count as being "after" their prefix (`1.2.3-really-final-this-time` counts as being released after `1.2.3`).
- GHC package numbers are comprised of any number of numeric segments separated by `.`, plus an optional (though deprecated) alphanumeric "version tag" separated by a `-`. I could not find any official documentation on ordering for the version tag; numeric segments are compared as numbers.
- npm uses semantic versioning so that is covered already.

After much more investigation than I'd care to admit, I have come up with the following compromise as the least-bad solution. First, obviously, the version string for stable/marketing versions is going to be "standard" semver, i.e. major.minor.patch, all numbers, which works, and sorts as expected, for all three schemes. For snapshot releases, we shall use the following (semver) format:

```
0.13.53-alpha.20200214.3025.ef5d32b7
```

where the components are, respectively:

- `0.13.53`: the expected version string of the next "stable" release.
- `alpha`: a marker that hopefully scares people enough.
- `20200214`: the date of the release commit, which _MUST_ be on master.
- `3025`: the number of commits in master up to the release commit (included). Because we have a linear, append-only master branch, this uniquely identifies the commit.
- `ef5d32b7ù : the first 8 characters of the release commit sha. This is not strictly speaking necessary, but makes it a lot more convenient to identify the commit.

The main downsides of this format are:

1. It is not a valid format for GHC packages. We do not publish GHC packages from the SDK (so far we have instead opted to release our Haskell code as separate packages entirely), so this should not be an issue. However, our SDK version currently leaks to `ghc-pkg` as the version string for the stdlib (and prim) packages. This PR addresses that by tweaking the compiler to remove the offending bits, so `ghc-pkg` would see the above version number as `0.13.53.20200214.3025`, which should be enough to uniquely identify it. Note that, as far as I could find out, this number would never be exposed to users.
2. It is rather long, which I think is good from a human perspective as it makes it more scary. However, I have been told that this may be long enough to cause issues on Windows by pushing us past the max path size limitation of that "OS". I suggest we try it and see what happens.

The upsides are:

- It clearly indicates it is an unstable release (`alpha`).
- It clearly indicates how old it is, by including the date.
- To humans, it is immediately obvious which version is "later" even if they have the same date, allowing us to release same-day patches if needed. (Note: that is, commits that were made on the same day; the release date itself is irrelevant here.)
- It contains the git sha so the commit built for that release is immediately obvious.
- It sorts correctly under all schemes (modulo the modification for GHC).

Alternatives I considered:

- Pander to GHC: 0.13.53-alpha-20200214-3025-ef5d32b7. This format would be accepted by all schemes, but will not sort as expected under semantic versioning (though Maven will be fine). I have no idea how it will sort under GHC.
- Not having any non-numeric component, e.g. `0.13.53.20200214.3025`. This is not valid semantic versioning and is therefore rejected by npm.
- Not having detailed info: just go with `0.13.53-snapshot`. This is what is generally done in the Java world, but we then lose track of what version is actually in use and I'm concerned about bug reports. This would also not let us publish to the main Maven repo (at least not more than once), as artifacts there are supposed to be immutable.
- No having a qualifier: `0.13.53-3025` would be acceptable to all three version formats. However, it would not clearly indicate to humans that it is not meant as a stable version, and would sort differently under semantic versioning (which counts it as a prerelease, i.e. before `0.13.53`) than under maven (which counts it as a patch, so after `0.13.53`).
- Just counting releases: `0.13.53-alpha.1`, where we just count the number of prereleases in-between `0.13.52` and the next. This is currently the fallback plan if Windows path length causes issues. It would be less convenient to map releases to commits, but it could still be done via querying the history of the `LATEST` file.

Release notes
=============

Release notes are a bit tricky. Because we want the ability to make snapshot releases, then later on promote them to stable releases, it follows that we want to build commits from the past. However, if we decide post-hoc that a commit is actually a good candidate for a release, there is no way that commit can have the appropriate release notes: it cannot know what version number it's getting, and, moreover, we now track changes in commit messages. And I do not think anyone wants to go back to the release notes file being a merge bottleneck.

But release notes need to be published to the releases blog upon releasing a stable version, and the docs website needs to be updated and include them.

The only sensible solution here is to pick up the release notes as of the commit that triggers the release. As the docs cron runs asynchronously, this means walking down the git history to find the relevant commit.

> Note: We could probably do away with the asynchronicity at this point. It was originally included to cover for the possibility of a release failing. If we are releasing commits from the past after they have been tested, this should not be an issue anymore. If the docs generation were part of the synchronous release step, it would have direct access to the correct release notes without having to walk down the git history.
>
> However, I think it is more prudent to keep this change as a future step, after we're confident the new release scheme does indeed produce much more reliable "stable" releases.

New release process
===================

Just like releases are currently controlled mostly by detecting changes to the `VERSION` file, the new process will be controlled by detecting changes to the `LATEST` file. The format of that file will include both the version string and the corresponding SHA.

Upon detecting a change to the `LATEST` file, CI will run the entire release process, just like it does now with the VERSION file. The main differences are:

1. Before running the release step, CI will checkout the commit specified in the LATEST file. This requires separating the release step from the build step, which in my opinion is cleaner anyway.
2. The `//:VERSION` Bazel target is replaced by a repository rule that gets the version to build from an environment variable, with a default of `0.0.0` to remain consistent with the current `daml-head` behaviour.

Some of the manual steps will need to be skipped for a snapshot release. See amended `release/RELEASE.md` in this commit for details.

The main caveat of this approach is that the official release will be a different binary from the corresponding snapshot. It will have been built from the same source, but with a different version string. This is somewhat mitigated by Bazel caching, meaning any build step that does not depend on the version string should use the cache and produce identical results. I do not think this can be avoided when our artifact includes its own version number.

I must note, though, that while going through the changes required after removing the `VERSION` file, I have been quite surprised at the sheer number of things that actually depend on the SDK version number. I believe we should look into reducing that over time.

CHANGELOG_BEGIN
CHANGELOG_END

gerolf-da commented:
Thank you for the detailed write up, @garyverhaegen-da.
I have two questions:

1) Looking a bit into the future, we will need to be able to perform a bug-fix release for a previously released version. Since the `master` branch will have moved on since then, we cannot run the bug-fix release off `master`, but will most likely have to run it from a branch. What are your thoughts on this topic? To be clear, I'm not asking that this PR supports this already, but I'm rather interested in your thinking.

2) Using the `alpha` marker might be conflicting with the initiative of ledger maturity/readiness levels that @meiersi-da has started (the levels being `Alpha`, `Beta`, `GA`). Even though release markers and labels for ledgers are on different levels of the stack, maybe we can avoid using the same term in both.

cocreature commented:
>     2\. The `//:VERSION` Bazel target is replaced by a repository rule that gets the version to build from an environment variable, with a default of `0.0.0` to remain consistent with the current `daml-head` behaviour.

Given the current state where a lot of things embed the SDK version number, won’t this mean that we basically end up rebuilding the world when we do a release? It also sounds like this might mean that we only catch issues caused by a different embedded SDK version (not that unlikely if you think of things like cross-SDK imports) once we do the release.

garyverhaegen-da commented:
> Thank you for the detailed write up, @garyverhaegen-da.
> I have two questions:
> 
> 1. Looking a bit into the future, we will need to be able to perform a bug-fix release for a previously released version. Since the `master` branch will have moved on since then, we cannot run the bug-fix release off `master`, but will most likely have to run it from a branch. What are your thoughts on this topic? To be clear, I'm not asking that this PR supports this already, but I'm rather interested in your thinking.

The main change in this PR, really, is that we now have the ability to take any random commit, slap a random string in an env var, and voilà, we have a new release with that string as its version number. There are a few minor things that depend on the master branch explicitly (`condition` field in the tarball & release steps of azure-pipelines, the way in which we fetch the release notes), but overall I'd say the changes in this PR are quite orthogonal to this.

My main concern, really, is version numbers: if you wanted to release a new "patch" on an existing release right now, say 0.13.45, you'd have to invent a version number for it. 0.13.45.1 would not be valid semver and npm will refuse it; 0.13.45-1 would not be a valid number for GHC, and would sort wrong on semver. You don't want to  call it an alpha release. My best advice there would be to stop changing the patch version on "normal" releases and keep it for that use-case.

I should take this opportunity to note that, in my opinion, it makes sense to have things that are released together in the same repo, but if things are released on different schedules, they should be in different repos, especially if you want to make any sort of guarantee about cross-compatibility between their releases.

> 2. Using the `alpha` marker might be conflicting with the initiative of ledger maturity/readiness levels that @meiersi-da has started (the levels being `Alpha`, `Beta`, `GA`). Even though release markers and labels for ledgers are on different levels of the stack, maybe we can avoid using the same term in both.

I was not aware of that initiative at all; thanks for bringing this up. I guess that leaves us with `snapshot` as the best alternative. I'm a little bit concerned that it's not as scary as "alpha" to a human reader, but I guess it will do.

garyverhaegen-da commented:
> > ```
> > 2\. The `//:VERSION` Bazel target is replaced by a repository rule that gets the version to build from an environment variable, with a default of `0.0.0` to remain consistent with the current `daml-head` behaviour.
> > ```
> 
> 
> Given the current state where a lot of things embed the SDK version number, won’t this mean that we basically end up rebuilding the world when we do a release?

Yes. There really is no way around that, though. As I noted, I was surprised by (and frankly a bit concerned at) the sheer number of things that currently depend on the version number. Currently on my machine making a "same" build with a different version number takes about 9 minutes and subsequently running the tests takes more than half an hour.

There are definitely things we can do to mitigate that (though I'd rather not add them to this PR), such as plain removing the inclusion if it's not strictly required, and otherwise moving the dependency "up" the chain as far as possible.

Note that this is a cost we are already paying, and this PR actually lowers it. In the existing system, the release commit changes the VERSION file, so the build of the release commit still needs to rebuild the same parts of the world. Moreover, after the release commit has been merged into master, every other PR has to rebuild or download (depending on cache population and propagation) the new artifacts from the cache. With these changes at least only the release PR has to "pay" this price, as everyone else keeps using the artifacts built against 0.0.0.

> It also sounds like this might mean that we only catch issues caused by a different embedded SDK version (not that unlikely if you think of things like cross-SDK imports) once we do the release.

I'm not sure I understand this, especially with respect to how it differs from the current situation. We currently always build with the version in VERSION, so there is also just one version in the test.

We could run the build & test with the version tag on the PR build rather than (only) after merging into master. Would that help?

garyverhaegen-da commented on LATEST:
```diff
@@ -0,0 +1 @@
+0f03751a35ffa8f05d52d4b1590aec9b366dadae 0.13.52-alpha.20200212.3104.626a89ee
```
`nightly` is not one of the magic qualifiers blessed by Maven, so they would sort after the corresponding release. `snapshot` would work, though.

garyverhaegen-da commented on bazel_tools/build_environment.bzl:
```diff
@@ -0,0 +1,42 @@
+""".format(
+                SDK_VERSION = semver,
+                NPM_VERSION = semver,
+                MVN_VERSION = "10" + semver,
```
I think we really should aim to fix this discrepancy before we reach 1.0, but I'll update this to parse the number and do +100 so we're safe.

garyverhaegen-da commented on ci/cron/src/Main.hs:
```diff
@@ -203,13 +203,29 @@ build_docs_folder path versions latest = do
             shell_ $ "cp -r " <> from <> " " <> to
         build version path = do
             shell_ $ "git checkout v" <> version
+            shell_ "git reset --hard"
```
This is to undo the checkout of the release notes file. Maybe I should move that to be closer so it's more obvious.

cocreature replied:
or just add a comment.

garyverhaegen-da replied:
I've restructured the code around this such that the `git checkout` and the corresponding `git reset` are part of the same `bracket` call, which should hopefully make it clear.

garyverhaegen-da commented on ci/cron/src/Main.hs:
```diff
@@ -203,13 +203,29 @@ build_docs_folder path versions latest = do
+            -- release notes still need to be taken from the release commit
+            -- (i.e. the one that changes the LATEST file, not the one being
+            -- pointed to).
+            if to_v version > to_v "0.13.53"
```
Yes, definitely.

garyverhaegen-da commented on ci/cron/src/Main.hs:
```diff
@@ -203,13 +203,29 @@ build_docs_folder path versions latest = do
+                -- need to find it by walking through the git history of the
+                -- LATEST file.
+                sha <- find_commit_for_version version
+                shell_ $ "git checkout " <> sha <> " -- docs/source/support/release-notes.rst"
```
The idea is that, when making the release, you have taken the release notes up to the source commit and put that in the release commit. So that should work out, as the release notes in the release commit are really as of the source commit and therefore cannot have references to things that came out after it.

cocreature replied:
I get the intention what I’m a bit worried about is that this is not what `bazel build //docs` checks. But I’m fine with this for now.

garyverhaegen-da replied:
I'm happy to consider any suggestion on how to improve this, but I can't see any way out of it myself.

garyverhaegen-da commented on compiler/daml-extension/BUILD.bazel:
```diff
@@ -91,19 +92,18 @@ genrule(
         set -euo pipefail
         TMP_DIR=$$(mktemp -d)
-        cleanup () { rm -rf $$TMP_DIR || return; }
+        cleanup () {{ rm -rf $$TMP_DIR || return; }}
```
Yes: because we add a call to .format at the end of that string, Bazel will take anything in squiggly brackets as a substitution; the double brackets prevent that.

garyverhaegen-da commented on compiler/damlc/lib/DA/Cli/Damlc.hs:
```diff
@@ -476,6 +476,24 @@ execLint inputFile opts =
+-- 0.13.51 -> release, no change
+-- 0.13.51-alpha.20200212.3024.04e6fa2c -> alpha release, change to
+         --                                0.13.51.20200212.3024 internally
+getVersion :: ProjectConfig -> Either ConfigError PackageSdkVersion
```
Will update, thanks.

garyverhaegen-da replied:
I've ended up completely restructuring this; I don't think the new approach introduces any confusing name.

garyverhaegen-da commented on rules_daml/daml.bzl:
```diff
@@ -136,15 +137,15 @@ def _daml_doctest_impl(ctx):
         package_name = ctx.attr.package_name,
         flags = " ".join(ctx.attr.flags),
-        version_file = ctx.file.version.path,
+        version = sdk_version,
```
I'm not sure, to be honest. I've tried to limit `ghc_version` to the minimum number of places where it is required, and things seem to build with the sdk version here. As far as I'm aware `ghc_version` is only needed for the one call to `ghc-pkg` inside the compiler; everything else should see the sdk version if possible.

garyverhaegen-da replied:
Turns out: yes, it should.

garyverhaegen-da commented on release/RELEASE.md:
```diff
@@ -1,18 +1,34 @@
+For the latter, there are extra steps marked as *[STABLE]* in the following
+instructions. You can safely skip those if you are making a prerelease,
+that is, one intended mostly for internal use and early testing, but which
+makes no promises about future compatibility.
```
I don't know. This is up for discussion in the (broader) team. This PR is about making it possible to have alpha releases; what level of guarantee we give about them is, in my opinion, quite orthogonal.

What I had in mind when writing that was things like changing the name of a new function in the API. I personally think people would be crazy to actually deploy an alpha/snapshot version in production where they have data they care about, but it's not up to me to decide, from a business perspective, what we do about it if it happens.

garyverhaegen-da commented on azure-pipelines.yml:
```diff
@@ -199,62 +229,150 @@ jobs:
+          set -euo pipefail
+          eval "$(./dev-env/bin/dade-assist)"
+
+          export DAML_SDK_RELEASE_VERSION=$(release_tag)
```
Makes sense. I'll see if I can add that to the PR build step, so we don't have to wait until we merge it to discover if that breaks things.

Spolier alert: I had not actually tried running `bazel test` with the env var set, and it does break things. I'll see if I can fix it.

garyverhaegen-da commented on release/src/Main.hs:
```diff
@@ -25,33 +25,26 @@ import Types
-      artifacts :: [Artifact (Maybe ArtifactLocation)] <- decodeFileThrow optsArtifacts
-
-      let targets = concatMap buildTargets artifacts
+      mvnArtifacts :: [Artifact (Maybe ArtifactLocation)] <- decodeFileThrow "release/artifacts.yaml"
```
I'll try. The refactorings were mostly to make my life easier in making the "real" changes so they may be a bit difficult to extract.

cocreature replied:
Thinking about it again, I’m fine with keeping this. I don’t want to cause even more work for you :slightly_smiling_face:

garyverhaegen-da commented on release/util.bzl:
```diff
@@ -54,18 +53,6 @@ def sdk_tarball(name, version):
           cp -t $$OUT/daml-libs $(location //triggers/daml:daml-trigger.dar)
           cp -t $$OUT/daml-libs $(location //daml-script/daml:daml-script.dar)
 
-          # Patch the SDK version in all bundled DAML libraries.
```
Right. In the current state we build all these libraries with the latest release version number, and then for a head build we need to go inside the packages and change the version number to `0.0.0` to match `daml-head`.

In other words, in the current setup we take a build that has been done with a different version and manually patch it to look like 0.0.0, whereas in this PR we use the same version number for both the sdk tarball build and the build of all the things it depends on, so there is no need for additional patching.

garyverhaegen-da commented on release/RELEASE.md:
```diff
@@ -122,10 +141,13 @@
+   announcements (public Slack, Twitter, etc.).
+   should appear automatically.
+
+1. *[STABLE]* Documentation is published automatically once the release is
```
Not really, no. We'd just need to figure out how/where to publish them. My current thinking (but this can easily be changed after this PR) is that prereleases should stay marked as prereleases on GitHub, and that currently makes them invisible to the docs cron.

garyverhaegen-da commented on release/RELEASE.md:
```diff
@@ -1,18 +1,34 @@
    Retrieve the new release notes using the command
 
-      `./unreleased.sh <revision range>`
+      `./unreleased.sh $LAST_VERSION..$SHA`
```
I'm not strongly decided on the matter. The instructions as written _would_ produce release notes for each alpha version, but the docs (and hence release notes) would only get published to docs.daml.com on stable releases. If followed, the current instructions would yield a release notes file with a series of entries for alpha releases and then an empty entry for the final release (provided it is just promoting the latest alpha to stable).

My current thinking (but this is very much up for debate) is that we should update the release notes file to have an entry for the latest alpha release. On a new alpha release, we should update that entry (i.e. not keep the old one if it was already an alpha), and when doing a stable release we should just change the title of the last entry.

garyverhaegen-da commented on release/RELEASE.md:
```diff
@@ -108,7 +127,7 @@
    1. Add `+` at the end of line 26 after `"Alice"` and verify that you get an error.
 
-1. If there are no issues, the release can be made public.
+1. *[STABLE]* If there are no issues, the release can be made public.
```
Right. That's also what currently triggers publication on docs.daml.com, so as written only stable releases would be published there.

garyverhaegen-da commented on bazel_tools/build_environment.bzl:
```diff
@@ -0,0 +1,42 @@
+
+    # Generates a simple Bazel file that just sets a bunch of Bazel variables,
+    # so they can be used in our main Bazel BUILD files.
+    semver = ctx.os.environ.get("DAML_SDK_RELEASE_VERSION", default = "0.0.0")
```
> Just to make sure I understand the implications of this correctly:
> 
> * Everything that currently depends on `//:VERSION` and embeds it in some form will default to embedding `0.0.0` instead.

Yes.

> * If I want to reproduce a release commit build, it is not sufficient to checkout the commit. I have to set the environment variable.

Yes.

> * In a release build everything that embeds the version number will get the the version number specified in `LATEST`.

With the exception of Maven artifacts, which get a +100, and some GHC-internal things which (for snapshot builds) get a slightly simplified version string.

This is not magic though, this is CI setting the env var based on the LATEST file of the release commit (the one triggering the release) while building the source commit (the one whose SHA is in the LATEST file in the release commit).

So if LATEST says `abc 0.1.2`, CI is doing

```
git checkout abc
DAML_SDK_RELEASE_VERSION=0.1.2 bazel build //...
```

> * `daml-sdk-head` will embed `0.0.0` whereas before it embedded things from `LATEST`.

`daml-sdk-head always had `0.0.0`. It used to take artifacts built with `VERSION` and then "manually" patching them to change their version number before embedding them in the tarball. Now it builds the tarball directly from artifacts built as `0.0.0`.

> My two main worries here are:
> 
> 1. Caching will get significantly worse for release builds.

Release builds were already specifically builds that changed the VERSION file, so this is not getting worse in any way that I can see.

> 2. If there are issues relating to the version number (I’m particularly thinking of `damlc`, `daml-stdlib` and cross-SDK dependencies here. I’m not _aware_ of anything that should go wrong but this is finicky), we will not catch them until we do a release build which actually changes the version number used in the build.

True, and thanks for raising. I will change the CI configuration to run those tests against the release PR (rather than not run them at all, which is what this PR currently does).

> 
> I don’t have a great answer for either of those. I’d feel a bit more comfortable about 2 if we ran tests on release builds with the version number set but that also means release builds will basically rebuild and retest the world (I believe both sandbox and damlc embed the version number). Still seems better than catching the issue due to manual testing.

I agree. Though I strongly suspect that if our tests break on a version change like that, it's going to be a false positive, i.e. a wrong assumption in the test rather than an error in the binary produced.

> 
> The only other idea that I have is to continue for now to embed the SDK version from `LATEST` if the environment variable is not set. That at least seems to reduce the number of moving parts compared to the current SDK release process and should help with 1 and partially with 2 although it doesn’t quite match up.

I don't think that solves anything, really. It would not change anything to 1; the only difference for 2 is that currently we do effectively run the test suite with the env var set on the release PR (so, again, adding this here would not make it worse, AFAICT).

hurryabit commented on daml-assistant/integration-tests/BUILD.bazel:
```diff
@@ -27,17 +27,16 @@ genrule(
           -Dpackaging=jar \
-          -Dversion=$$VERSION
-      }
+          -Dversion={}
```
Unnamed placeholders scare me. (Maybe I haven't used enough C-style `printf` in my life?)
```suggestion
          -Dversion={mvn_version}
```
There are a few more instances of this elsewhere.

hurryabit commented on daml-script/daml/BUILD.bazel:
```diff
@@ -3,29 +3,30 @@
 name: daml-script
 source: daml
-version: $$(cat $(location //:VERSION))
+version: {}
```
Two unnamed placeholders in the same interpolation scare me even more.
```suggestion
version: {ghc_version}
```

rohanjr commented on release/RELEASE.md:
```diff
@@ -122,10 +141,13 @@
-   however it takes up to an hour to complete.
+1. *[STABLE]* Coordinate with product (& marketing) for the relevant public
+   announcements (public Slack, Twitter, etc.).
+   should appear automatically.
```
Delete this line?

garyverhaegen-da commented:
I have updated the PR to what I now believe to be a fully working version that contains all the feedback I have received. Highlights of the changes:
- I can now confirm that running `bazel build` and `bazel test` on this PR with the env var set to the new version format works (at least as of just before my latest rebase; I'm running the new one now).
- The PR has also been rebased and fixed up (some changes to //BUILD and the new repl work did conflict)
- The new CI job that checks whether the current PR is a release PR now also fails if the version number in LATEST is malformed. (It also now checks the entire PR instead of just the latest commit when listing the files that have changed.)
- `alpha` has been renamed to `snapshot` throughout
- The Maven number computation now does a proper arithmetic +100 on the major version number.
- The release program now uses the `bracket` function to explicitly link the `git checkout` of the release notes and the corresponding `git reset --hard`.
- For release PRs, the build jobs (Linux, macOS, windows & perf) will now checkout the target release SHA and run with the environment variable set to the target release tag. This should ensure that, once the PR is merged, the build can only fail for transient reasons.
- This PR no longer introduces any unnamed string substitution, and actually removes some existing ones.
- The release instructions in release/RELEASE.md have been updated to match the decisions made last week (i.e. no docs or release notes for alpha releases; team lead approval required for stable release).

garyverhaegen-da commented:
Edits:
- Clean up isStableVersion.
- Add back the condition that we only upload tarballs from master. Somehow it had been lost in the restructuring of the release steps.
